src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs

```diff
@@ -329,41 +318,98 @@ static async Task<HttpResponseMessage> Impl(HttpRequestMessage request, Cancella
         }
     }
 
-    internal sealed class ReadableStreamPullState
+    internal sealed class WasmHttpWriteStream : Stream
     {
-        private readonly Stream _stream;
-        private readonly CancellationToken _cancellationToken;
-        private readonly byte[] _buffer;
+        private readonly JSObject _requestStream;
 
-        public ReadableStreamPullState(Stream stream, CancellationToken cancellationToken)
+        public WasmHttpWriteStream(JSObject requestStream)
         {
-            ArgumentNullException.ThrowIfNull(stream);
+            ArgumentNullException.ThrowIfNull(requestStream);
 
-            _stream = stream;
-            _cancellationToken = cancellationToken;
-            _buffer = new byte[65536];
+            _requestStream = requestStream;
         }
 
-        public async Task PullAsync()
+        public static async Task<JSObject> CopyToAsync(JSObject requestStream, HttpContent content, CancellationToken cancellationToken)
         {
-            try
-            {
-                int length = await _stream.ReadAsync(_buffer, _cancellationToken).ConfigureAwait(true);
-                ReadableStreamControllerEnqueueUnsafe(this, _buffer, length);
-            }
-            catch (Exception ex)
+            using (WasmHttpWriteStream stream = new WasmHttpWriteStream(requestStream))
             {
-                BrowserHttpInterop.ReadableStreamControllerError(this, ex);
+                try
+                {
+                    await content.CopyToAsync(stream, cancellationToken).ConfigureAwait(true);
+                    return await BrowserHttpInterop.RequestStreamClose(requestStream).ConfigureAwait(true);
+                }
+                catch (Exception ex)
+                {
+                    await BrowserHttpInterop.RequestStreamClose(requestStream, ex).ConfigureAwait(true);
+                    throw;
+                }
             }
         }
 
-        private static unsafe void ReadableStreamControllerEnqueueUnsafe(object pullState, byte[] buffer, int length)
+        private async Task WriteAsyncCore(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
         {
-            fixed (byte* ptr = buffer)
+            cancellationToken.ThrowIfCancellationRequested();
+            using (Buffers.MemoryHandle handle = buffer.Pin())
             {
-                BrowserHttpInterop.ReadableStreamControllerEnqueue(pullState, (nint)ptr, length);
+                await RequestStreamWriteUnsafe(_requestStream, buffer, handle).ConfigureAwait(true);
             }
+
+            static unsafe Task RequestStreamWriteUnsafe(JSObject requestStream, ReadOnlyMemory<byte> buffer, Buffers.MemoryHandle handle)
```

Comment:
Please make it cancellable promise. The pattern is already in the previous code. Or I can try to help :)

Reply:
When this propagates cancelation exception, or if this was aborted,
how do we propagate that error to the caller of the top level `.SendAsync`, does that work fine ?

Reply:
I am planning to do it like this https://github.com/dotnet/runtime/pull/91699#discussion_r1319082854 so it should be automatically cancellable since fetch is cancellable.